cyto_dl/api/cyto_dl_model/cyto_dl_base_model.py

```diff
@@ -0,0 +1,135 @@
+
+
+class CytoDLBaseModel(ABC):
+    """A CytoDLBaseModel is used to configure, train, and run predictions on a cyto-dl model."""
+
+    def __init__(self, config_filepath: Optional[Path] = None):
+        """
+        :param config_filepath: path to a .yaml config file that will be used as the basis
+        for this CytoDLBaseModel (must be generated by the CytoDLBaseModel subclass that wants
+        to use it). If None, a default config will be used instead.
+        """
+        self._cfg: DictConfig = (
+            OmegaConf.load(config_filepath) if config_filepath else self._generate_default_config()
+        )
+
+    def _generate_default_config(self) -> DictConfig:
+        cfg_dir: Path = (
+            pyrootutils.find_root(search_from=__file__, indicator=("pyproject.toml", "README.md"))
+            / "configs"
+        )
+        GlobalHydra.instance().clear()
+        with initialize_config_dir(version_base="1.2", config_dir=str(cfg_dir)):
+            # the overrides for 'paths.' are necessary to make later overrides work (for some reason I don't fully understand)
+            # I'd prefer to do this in experiment configs instead of the overrides here
+            cfg: DictConfig = compose(
+                config_name="train.yaml",  # only using train.yaml after conversation w/ Benji
+                return_hydra_config=True,
+                overrides=[
+                    f"experiment=im2im/{self._get_experiment_type().name}",
+                    "paths.output_dir=PLACEHOLDER",
+                    "paths.work_dir=PLACEHOLDER",
+                ],
+            )
+        with open_dict(cfg):
+            del cfg["hydra"]
+            cfg.extras.enforce_tags = False
+            cfg.extras.print_config = False
+        return cfg
+
+    @abstractmethod
+    def _get_experiment_type(self) -> ExperimentType:
+        """Return experiment type for this config (e.g. segmentation_plugin, gan, etc)"""
+        pass
+
+    def _key_exists(self, k: str) -> bool:
+        keys: List[str] = k.split(".")
+        curr_dict: DictConfig = self._cfg
+        while keys:
+            key: str = keys.pop(0)
+            if not key in curr_dict:
+                return False
+            curr_dict = curr_dict[key]
+        return True
+
+    def _set_cfg(self, k: str, v: Any) -> None:
+        if not self._key_exists(k):
+            raise KeyError(f"{k} not found in config dict")
+        OmegaConf.update(self._cfg, k, v)
+
+    def _get_cfg(self, k: str) -> Any:
+        if not self._key_exists(k):
+            raise KeyError(f"{k} not found in config dict")
+        return OmegaConf.select(self._cfg, k)
+
+    def _set_training_config(self, train: bool):
+        self._set_cfg("train", train)
+        self._set_cfg("test", train)
+        # afaik, task_name isn't used outside of template_utils.py - do we need to support this?
+        self._set_cfg("task_name", "train" if train else "predict")
+
+    @abstractmethod
```

Comment:
I find it helpful to group all the abstract methods together at the top or bottom.

Reply:
Good call

Reply:
[bbaeeb](https://github.com/AllenCellModeling/cyto-dl/pull/361/commits/33bee7ac8833c56b8aed77f5a21c39bc15bbaeeb)